internal/receiver/smartagentreceiver/receiver.go

```diff
@@ -135,5 +145,63 @@ func (r *Receiver) createMonitor(monitorType string) (interface{}, error) {
 		output.AddExtraDimension(k, v)
 	}
 
+	collectdLock.Lock()
+	defer collectdLock.Unlock()
+	// Note, that this receiver has to configure main collectd even for collectd/custom monitor,
+	// despite the fact that, that monitor stands up its own instance of collectd to prevent this
+	// panic "Main collectd instance should not be accessed before being configured".
+	if r.config.monitorConfig.MonitorConfigCore().IsCollectdBased() && !configuredCollectd {
+		r.setUpCollectdConfig(extensions)
+		if err := collectd.ConfigureMainCollectd(r.getCollectdConfig()); err != nil {
+			return nil, err
+		}
+		configuredCollectd = true
+	}
```

Comment:
Can we extract this as a func? I think it will help to understand why we need this, which is not quite clear. Does this initialize the a singleton collectd instance? (I read the comment, but it is still not completely clear to me what are we doing here).
From what I see the mutex+bool flag is used to execute this once, right? Maybe use once.Do instead to make it clearer? (https://golang.org/pkg/sync/#Once)

Reply:
Yes,  `collectd.ConfigureMainCollectd` should only be called once. This call sets up the collectd singleton. Updated to use `sync.Once`